app/src/main/java/cm/aptoide/pt/billing/payment/SharedPreferencesPaymentServiceSelector.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2017.
+ * Modified by Marcelo Benites on 17/02/2017.
+ */
+
+package cm.aptoide.pt.billing.payment;
+
+import android.content.SharedPreferences;
+import cm.aptoide.pt.billing.payment.PaymentService;
+import cm.aptoide.pt.billing.payment.PaymentServiceSelector;
+import java.util.List;
+import rx.Completable;
+import rx.Observable;
+import rx.Single;
+import rx.schedulers.Schedulers;
+
+public class SharedPreferencesPaymentServiceSelector implements PaymentServiceSelector {
+
+  private static final String SELECTED_SERVICE_TYPE = "SELECTED_SERVICE_TYPE";
+  private final String defaultServiceType;
+  private final SharedPreferences preferences;
+
+  public SharedPreferencesPaymentServiceSelector(String defaultServiceType,
+      SharedPreferences preferences) {
+    this.defaultServiceType = defaultServiceType;
+    this.preferences = preferences;
+  }
+
+  @Override public Single<PaymentService> selectedService(List<PaymentService> services) {
+    return getSelectedServiceName().flatMap(
+        serviceName -> getService(services, serviceName).switchIfEmpty(
+            getService(services, defaultServiceType))
+            .switchIfEmpty(Observable.just(services.get(0)))
+            .first()
+            .toSingle());
+  }
+
+  @Override public Completable selectService(PaymentService service) {
+    return Completable.fromAction(() -> preferences.edit()
+        .putString(SELECTED_SERVICE_TYPE, service.getType())
+        .commit())
```

Comment:
Consider using apply() instead

Reply:
The observable is subscribed on I/O scheduler therefore I can execute the commit synchronously. In fact I have to make sure it finishes the operation before the action returns. Using apply I could end up with an inconsistent state.